File: backend/storage/src/main/java/com/intetm/model/Word.java
package com.intetm.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.google.gson.annotations.Expose;

import javax.persistence.*;
import java.util.Date;
import java.util.List;

@Entity
@Table(name = "WORD")
public class Word {

    @Id
    @Expose
    @Column(name = "id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    @Column
    @Expose
    public String text;

    @JsonIgnore
    @Column
    public String ltext;

    @ManyToOne
    @JoinColumn(name = "language", referencedColumnName = "id")
    public Language language;

    @JsonIgnore
    @ManyToOne
    @JoinColumn(name = "dictionary", referencedColumnName = "id")
    public Dictionary dictionary;

    @JsonIgnore
    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "from")
    public List<WordLink> from;

    @JsonIgnore
    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "to")
    public List<WordLink> to;
    @Column(name = "count_attempts")
    @Expose
    public int countAttempts;
    @Column(name = "count_success")
    @Expose
    public int countSuccess;
    @Column(name = "count_fail")
    @Expose
    public int countFail;
    @Column(name = "created")
    @Expose
    public Date created;
    @Column(name = "last_attempt")
    @Expose
    public Date lastAttempt;
    @Column(name = "last_success")
    @Expose
    public Date lastSuccess;

    public Word() {
    }

    public Word(Long id) {
        this.id = id;
    }

}

File: frontend/src/app/trainer/service/entity/word.ts
import {IdGenerator} from "../send/IdGenerator";

export class WordLink {
    id: number;
    from: Word[];
    to: Word[];
    transportId: number;

    constructor(form: Word[], to: Word[]) {
        this.id = null;
        this.from = form;
        this.to = to;
        this.transportId = IdGenerator.generateTransportId();
    }
}

export class Word {
    id: number;
    transportId: number;
    public text: string;
    created: Date;
    language: number;
    lastAttempt: Date;
    lastSuccess: Date;
    countAttempts: number = 0;
    countSuccess: number = 0;
    countFail: number = 0;

    constructor(text: string) {
        this.id = null;
        this.text = text;
        this.created = new Date();
        this.transportId = IdGenerator.generateTransportId();
    }

    static update(w1: Word, w2: Word) {
        w1.countAttempts = w2.countAttempts;
        w1.countFail = w2.countFail;
        w1.countSuccess = w2.countSuccess;
    }
}
